File: practica2/c++/programa.cc
#include <stack>
#include <iomanip>
#include <iostream>
#include "programa.h"

using namespace std;

programa::programa(int _inst)
	: pc(0), numInst(_inst)
{
	instrucciones = new InstructionPtr[numInst];
}


programa::~programa()
{
	for(int i = 0; i< numInst; i++)
	{
		delete instrucciones[i];
	}
}

void programa::ejecutar()
{
	stack<int32_t> pila;
	while(pc < numInst)
	{
		instrucciones[pc]->ejecutar(pila, pc);
	}
}

void programa::listar()
{
	for(int i = 0; i< numInst; i++)
	{
		cout << setw(2) << i <<  "   " << instrucciones[i]->nombre() << endl;
	}
}

File: practica2/java/Programa.java
import java.util.Stack;
/**
 * programa
 */
public class Programa
{
	protected int pc;
	protected int numInst;
	protected Inst[] instrucciones;

    public Programa(int _inst) {
    pc = 0;
    numInst = _inst;
	instrucciones = new Inst[numInst];
    }


    public void ejecutar()
    {
        Stack<Integer> pila = new Stack<Integer>();
        pc = 0;
	    while(pc < numInst)
	    {
           // System.out.printf("%d\n", pc);
		    pc = instrucciones[pc].ejecutar(pila, pc);
           // System.out.printf("%d\n", pc);
            /*
            if(pila.empty()) {
                System.out.print("La pila está vacía\n");
            }
            else {
                System.out.printf("Cima de la pila: %d\n", pila.peek());
            }
            */
	    }
}

    public void listar()
    {
	    for(int i = 0; i< numInst; i++)
	    {
            System.out.printf("%2d   %s\n", i, instrucciones[i].nombre());
	    }
    }
}
